### DISCUSSION TITLE
Use pre-releases

### DISCUSSION THREAD

**Discussion by VVishion** [bevyengine/bevy#2782] [Ideas]
See https://doc.rust-lang.org/cargo/reference/resolver.html#pre-releases for pre-releases.
Use pre-releases for bevy to allow plugin maintainers to target the current, pre-release bevy version.

This enables plugin maintainers to target the plugin at bevy between two releases. Therefore plugins must not be locally modified by everyone trying to run it on the current bevy pre-release.

Results in less friction for the individual using up-to-date bevy if someone already published the changes needed for targeting the pre-release.

Things unclear:
* What should the plugin exactly depend on? pre-release bevy on crates.io, github? a commit?
* We dont want to select pre-release candidates. Does crates.io allow this automation of publishing from github?

**Comment by bjorn3** [comment]
> Does crates.io allow this automation of publishing from github?

Yes, but publishing at every commit is a bad idea due to the amount of resources it consumes at crates.io.

> Use pre-releases for bevy to allow plugin maintainers to target the current, pre-release bevy version.

All pre-releases are semver compatible with each other. Plugins will likely only work with a subset of the pre-releases though. Also there may not be a subset which works for all plugins you want.

> What should the plugin exactly depend on? pre-release bevy on crates.io, github? a commit?

I think it should just specify the git repo of bevy instead of a pre-release from crates.io.

**Comment by Nilirad** [comment]
I think it's better to have the upcoming big bevy features into a separate branch kept in sync with `main`, just like `pipelined_rendering`, such that a plugin maintainer can use it to keep up with the next changes if they want to do it beforehand.

It doesn't require additional effort or complicating the release process.

Edit: Also, if a plugin maintainer wants to keep up with the latest Bevy release as soon as possible, one of their branches (possibly the default branch) should really depend on Bevy's `main` branch. Otherwise it's a hell to fix everything in time given the huge amount of changes that happened between versions.

**Comment by VVishion** [comment]
I think the ability to target big upcoming features is important.
But they should be kept off `main`, if pretty much unstable. `main` should be kept in a useable state, especially in bevys infancy state.
Plugins should target the bevy feature branch on git, then.
Are plugins already doing this?
And if this is a working and desired approach it would be nice to see it largely documented somewhere.

**Reply by Nilirad** [reply]
> Plugins should target the bevy feature branch on git, then.
Are plugins already doing this?

Yes, I'm planning to do something similar with `bevy_prototype_lyon`. This is my policy:

- `master` depends on Bevy `main`
- other branches depend on Bevy feature branches. Those branches are kept in sync with `master` to minimize merge friction.

I recognized that mirroring the branch structure of Bevy is the simplest approach, at least for my case.
